File: 2024/src/main.jl
# Packages
using ArgParse
include("julia/download.jl")

# Function to parse arguments
function parse_input()
    parser = ArgParseSettings(description="Advent of Code 2024 in Julia")

    @add_arg_table! parser begin
        "day"
        arg_type = Int
        help = "Day to solve"

        "--year"
        arg_type = Int
        default = 2024
        help = "Year of Advent of Code"
    end

    return parse_args(ARGS, parser)
end

# Function to dynamically load and run the solve function
function main()
    
    # Parse arguments
    args = parse_input()
    day = args["day"]
    year = args["year"]


    # Download data (implement this function as needed)
    download(day, year)

    # Build module file path
    module_path = joinpath(@__DIR__, "julia", "day$day.jl")

    # Check if the file exists
    if !isfile(module_path)
        println("Julia code for day $day not found.")
        return
    end

    # Dynamically include the module
    include(module_path)

    # Get the module name dynamically
    module_name = Symbol("Day$day")

    # Check if the module is loaded
    if !haskey(Base.loaded_modules, module_name)
        println("Julia code for day $day not found.")
        return
    end

    # Access the module
    mod = Base.require(module_name)

    # Check if the solve function exists
    if !hasproperty(mod, :solve)
        println("Julia code for day $day does not have a solve function.")
        return
    end

    # Call the solve function
    solve = getproperty(mod, :solve)
    solve()
end

# Run the main function
main()

File: 2024/src/main.py
# Import packages
import argparse
import importlib
from python.download import download


def main():
    # Create parser
    parser = argparse.ArgumentParser(description='Advent of Code 2024 in Python')

    # Add arguments
    parser.add_argument('day', type=int, help='Day to solve')
    parser.add_argument('--year', type=int, default=2024, help='Year of Advent of Code')

    # Parse arguments
    args = parser.parse_args()

    # Download data
    download(args.day, args.year)

    # Dynamically import the solution module
    try:
        module_name = f'py.day{args.day}'
        solution_module = importlib.import_module(module_name)

        # Call the solve() function
        if hasattr(solution_module, 'solve'):
            solution_module.solve()
        else:
            print(f'Python solve not implemented for day {args.day}.')
    except ModuleNotFoundError:
        print(f'Python code for day {args.day} not found.')
    except Exception as e:
        print(f'An error occurred for python code for day {args.day}, {e}.')


if __name__ == '__main__':
    main()
